connections/muxer-sel-in-sec-handshake.md

```diff
@@ -0,0 +1,266 @@
+
+If the client and server agree upon the common muxer to be used, then the
+result of the muxer selection is a muxer code represented by the selected
+protocol name. If no agreement is reached upon by the client and server
+then an empty muxer code is returned and the connection upgrade process
+MUST fall back to the multistream-selection protocol to negotiate the muxer.
+
+
+### Muxer selection over TLS
+
+When the security protocol selected by the upgrader is TLS, the [ALPN]
+extesion of TLS handshake is used to select the muxer.
+
+   With ALPN, the client sends the list of supported application
+   protocols as part of the TLS ClientHello message.  The server chooses
+   a protocol and sends the selected protocol as part of the TLS
+   ServerHello message.  The application protocol negotiation can thus
+   be accomplished within the TLS handshake, without adding network
+   round-trips, and allows the server to associate a different
+   certificate with each application protocol, if desired.
+
+For the purpose of muxer selection, the types of muxers are coded as protocol
+names in the form of a list of strings, and inserted in the ALPN extension
+field. An example list as following:
+
+    ["/yamux/1.0.0", "/mplex/6.7.0", "libp2p"]
+
+The muxer list is ordered by preference, with the most prefered muxer at
+the beginning. The "libp2p" protocol code MUST always be the last item in the
+muxer list . See [#tls-case] for details on the special "libp2p" protocol code.
+
+The server chooses the supported protocol by going through its prefered
+protocol list and searchs if the protocol is supported by the client too. if no
+mutually supported protcol is found the TLS handshake will fail.
+
+If the selected item from the muxer list is "libp2p" then the muxer selection
+process returns an empty result, and the multistream-selection protocol MUST be
+run to negotiate the muxer.
+
+
+### Muxer selection over Noise
+
+The libp2p Noise implementation allows the Noise handshake process to carry
+early data. [Noise-Early-Data] is carried in the second and third message of
+the XX handshake pattern as illustrated in the following message sequence chart.
+The second message carries early data in the form of a list of muxers supported
+by the responder, ordered by preference. The initiator sends its supported
+muxer list in the third message to the responder. After the Noise handshake
+process is fully done, the initiator and responder will both process the
+received eraly data and select the muxer to be used, they both iterate through
+the initiator's prefered muxer list in order, and if any muxer is also
+supported by the responder, that muxer is selected. If no mutually supported
+muxer is found, the muxer selection process MUST fall back to multistream
+-selection protocol.
+
+Example: Noise handshake between peers that have a mutually supported muxer.
+    Responder supports: ["yamux/1.0.0", "/mplex/6.7.0"]
+    Initiator supports: ["yamux/1.0.0"]
```

Comment:
Better: use an example that has [ mplex, yamux ], so that the preference becomes clear.

Reply:
done